components/automate-ui/src/app/modules/infra-proxy/environment-details/environment-details.component.ts

```diff
@@ -0,0 +1,139 @@
+import { Component, OnInit, OnDestroy, EventEmitter, ViewChild } from '@angular/core';
+import { Router } from '@angular/router';
+import { Store } from '@ngrx/store';
+import { combineLatest, Subject } from 'rxjs';
+import { NgrxStateAtom } from 'app/ngrx.reducers';
+import { LayoutFacadeService, Sidebar } from 'app/entities/layout/layout.facade';
+import { routeParams, routeURL } from 'app/route.selectors';
+import { filter, pluck, takeUntil } from 'rxjs/operators';
+import { identity } from 'lodash/fp';
+import { environmentFromRoute } from 'app/entities/environments/environment-details.selectors';
+import { GetEnvironment } from 'app/entities/environments/environment.action';
+import {
+  Environment,
+  CookbookVersion,
+  EnvironmentAttributes
+} from 'app/entities/environments/environment.model';
+import { JsonTreeTableComponent as JsonTreeTable } from './../json-tree-table/json-tree-table.component';
+
+export type EnvironmentTabName = 'cookbookConstraints' | 'attributes';
+
+@Component({
+  selector: 'app-environment-details',
+  templateUrl: './environment-details.component.html',
+  styleUrls: ['./environment-details.component.scss']
+})
+
+export class EnvironmentDetailsComponent implements OnInit, OnDestroy {
+  public environment: Environment;
+  public cookbookVersions: CookbookVersion[];
+  public tabValue: EnvironmentTabName = 'cookbookConstraints';
+  public url: string;
+  public conflictErrorEvent = new EventEmitter<boolean>();
+  public modalType: string;
+  public serverId;
+  public orgId;
+  public name;
```

Comment:
some types here would be great so we know exactly what they are.

Reply:
I have added :)